### PR TITLE
Changes to output Restart data for the WSEGAICD keyword (Eclipse compatible restart file)

### PR DESCRIPTION
**Pull request by jalvestad**
This pullrequest contains code changes to provide output of Restart data to the Eclipse compatible restart file to support the WSEGAICD keyword. 

I acknowledge Torbjørn Skille for providing the main input for the meaning of the restart-variables and Bård Skaflestad for valueable advice. 

This includes output to the restart tables: ISEG and RSEG as a part of the multisegment well restart output. The details of the 

Unit tests have been added for this keyword. In addition the test data for alle unit tests of the multisegment wells have been modified such that it runs with both flow and Eclipse and a direct comparison between the two datasets is possible. This enables an extra quality check for the correctness of the output restart data. The input data for the unit tests have been moved to a separate file in opm-common/tests. 

The PR passess unit tests and also the regression tests (make check) - as far as I have run the tests locally.

This PR is ready for review and merge from my side after review and modifications based on comments from reviewers.

### CODE REVIEW

**Comment by bska** [comment]
jenkins build this please

File: src/opm/output/eclipse/AggregateMSWData.cpp [hunk 1]
```diff
@@ -438,6 +438,20 @@ namespace {
...
+        template <class ISegArray>
+        void assignAICDCharacteristics(const Opm::Segment& segment,
+                                            const std::size_t   baseIndex,
+                                            ISegArray&          iSeg)
+        {
+            using Ix = ::Opm::RestartIO::Helpers::
+                VectorItems::ISeg::index;
+
+            const auto& aicd = segment.autoICD();
+            iSeg[baseIndex + Ix::SegmentType] = -8;
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateMSWData.cpp:450]: Magic number alarm!

File: src/opm/output/eclipse/AggregateMSWData.cpp [hunk 2]
```diff
@@ -610,6 +626,72 @@ namespace {
...
+        {
+            using Ix = ::Opm::RestartIO::Helpers::VectorItems::RSeg::index;
+            using M  = ::Opm::UnitSystem::measure;
+
+            const auto& aicd = segment.autoICD();
+
+            rSeg[baseIndex + Ix::DeviceBaseStrength] =
+                usys.from_si(M::aicd_strength, aicd.strength());
+
+            rSeg[baseIndex + Ix::ScalingFactor] = ((aicd.methodFlowScaling() == 1) ||
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateMSWData.cpp:643]: Please simplify - this is mouthful to read.

File: src/opm/output/eclipse/AggregateMSWData.cpp [hunk 3]
```diff
@@ -610,6 +626,72 @@ namespace {
...
+                aicd.maxViscosityRatio();
+
+            rSeg[baseIndex + Ix::FlowRateExponent] =
+                aicd.flowRateExponent();
+
+            rSeg[baseIndex + Ix::ViscFuncExponent] =
+                aicd.viscExponent();
+
+            rSeg[baseIndex + Ix::MaxValidFlowRate] =
+            (aicd.maxAbsoluteRate() == std::numeric_limits<double>::max()) ?
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateMSWData.cpp:668]: Hmm floating point `==` comparison?

File: src/opm/output/eclipse/AggregateMSWData.cpp [hunk 4]
```diff
@@ -610,6 +626,72 @@ namespace {
...
+               -2.e+20 : usys.from_si(M::geometric_volume_rate, aicd.maxAbsoluteRate()) ;
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateMSWData.cpp:669]: Magic number alert!

File: src/opm/parser/eclipse/Units/UnitSystem.cpp [hunk 5]
```diff
@@ -88,6 +88,7 @@ namespace {
         0.0,
         0.0,
         0.0,
+        0.0,
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/Units/UnitSystem.cpp:91]: New unit - typically something which should go in a separate PR as soon as the need is identified.

File: src/opm/output/eclipse/AggregateMSWData.cpp (hunk 1 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateMSWData.cpp:450]: I agree that we do not know why the segment type value is -8, however, we know that the value for an AICD segment is this value. I have added a comment about that to explain.

File: src/opm/output/eclipse/AggregateMSWData.cpp (hunk 2 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateMSWData.cpp:643]: I have added some explanatory comments for the logic test used to explain, is that OK?

File: src/opm/output/eclipse/AggregateMSWData.cpp (hunk 3 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateMSWData.cpp:668]: The reason for this test is that Eclipse uses a different value (-2.e+20) for the case when no maximum limit has been assigned. Flow then uses this maximum numeric limit value, which I have used for comparison - I do not know a different way to identify this special case. Please advice a different approach, I am happy to change the code!

File: src/opm/output/eclipse/AggregateMSWData.cpp (hunk 4 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateMSWData.cpp:669]: Se previous comment, Eclipse often uses these special numeric values for various default cases, this is one of the values used.

File: src/opm/parser/eclipse/Units/UnitSystem.cpp (hunk 5 above)
Comment (jalvestad) [reply src/opm/parser/eclipse/Units/UnitSystem.cpp:91]: I see the point, however, on the other hand this change is strongly connected the need to use this new unit conversion in this PR and the combination also gives a test that the code for the conversion actually works OK. Hence I hope they can be combined in this case.

Comment (joakim-hove) [reply src/opm/parser/eclipse/Units/UnitSystem.cpp:91]: My point is just that these changes could have been sent out at an earlier stage in the development process. That would have made the main PR somewhat smaller - and that is always good. No reason to split it off now.

File: src/opm/output/eclipse/AggregateMSWData.cpp (hunk 1 above)
Comment (joakim-hove) [reply src/opm/output/eclipse/AggregateMSWData.cpp:450]: Sure - I understand that it actually is a magic number - in the true sense of the word; my point is just that it should be named as a named constant - e.g.
```
SegmentypeAICD = 8;
```
in the `M` namespace.

**Comment by jalvestad** [comment]
I have looked at the failures in the previous Jenkins run and all the differences are OK and consistent with the new restart data output to ISEG and RSEG as far as I can judge - relates to the use of WSEGAICD data.

File: src/opm/output/eclipse/AggregateMSWData.cpp (hunk 2 above)
Comment (joakim-hove) [reply src/opm/output/eclipse/AggregateMSWData.cpp:650]: I did not really find a simple way myself either :-(

File: src/opm/output/eclipse/AggregateMSWData.cpp (hunk 3 above)
Comment (joakim-hove) [reply src/opm/output/eclipse/AggregateMSWData.cpp:675]: My main point was that value `-2e+20` should have a name - I guess the value `1e+20` is defined as a max somewhere? For the rate my suggestion would be to change[1] the `SICD::maxAbsoluteRate()` to return a `std::optional<double>` - then the whole limit value can be avoided:
```
const auto& max_rate  = aicd.maxAbsoluteRate();
if (max_rate.has_value())
      rSeg[baseIndex + Ix::MaxValidFlowRate] = usys.from_si( M::geometric_volume_rate, max_rate.valuie());
else 
     rSeg[] = -2 * EclipseInfinity;
```

File: src/opm/output/eclipse/AggregateMSWData.cpp (hunk 1 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateMSWData.cpp:451]: Good remark: I will use segment.ecl_type_id(), should have used that in the first place!!

File: src/opm/output/eclipse/AggregateMSWData.cpp (hunk 3 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateMSWData.cpp]: Thanks for good suggestion, I have changed the code as suggested!

**Comment by jalvestad** [comment]
Thanks again for good comments, I think I have changed the code as suggested by reviewer and consider the corrections completed from my side.

**Comment by joakim-hove** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
jenkins build this opm-tests=499 please

**Comment by joakim-hove** [comment]
> Any reservation towards merging this

No - in she goes!
